src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -1,81 +1,131 @@
+import io.iohk.ethereum.ledger.BlockExecutionError
 import io.iohk.ethereum.ledger.BlockExecutionError.MPTError
 import io.iohk.ethereum.ledger.BlockExecutionError.ValidationBeforeExecError
+import io.iohk.ethereum.ledger.BlockExecutionSuccess
+import io.iohk.ethereum.ledger.BlockMetrics
+import io.iohk.ethereum.ledger.BlockQueue
 import io.iohk.ethereum.ledger.BlockQueue.Leaf
+import io.iohk.ethereum.ledger.BlockValidation
 import io.iohk.ethereum.mpt.MerklePatriciaTrie.MissingNodeException
 import io.iohk.ethereum.utils.BlockchainConfig
 import io.iohk.ethereum.utils.ByteStringUtils
 import io.iohk.ethereum.utils.Logger
 
-class BlockImport(
+class ConsensusImpl(
     blockchain: BlockchainImpl,
     blockchainReader: BlockchainReader,
     blockchainWriter: BlockchainWriter,
     blockQueue: BlockQueue,
     blockValidation: BlockValidation,
-    private[ledger] val blockExecution: BlockExecution,
+    blockExecution: BlockExecution,
     validationScheduler: Scheduler // Can't be implicit because of importToTop method and ambiguous of Scheduler
-) extends Logger {
+) extends Consensus
+    with Logger {
 
   /** Tries to import the block as the new best block in the chain or enqueue it for later processing.
     *
-    * @param block                 block to be imported
-    * @param blockExecutionContext threadPool on which the execution should be run
+    * @param block                   block to be imported
+    * @param blockExecutionScheduler threadPool on which the execution should be run
+    * @param blockchainConfig        blockchain configuration
     * @return One of:
-    *         - [[io.iohk.ethereum.ledger.BlockImportedToTop]] - if the block was added as the new best block
-    *         - [[io.iohk.ethereum.ledger.BlockEnqueued]] - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
-    *         - [[io.iohk.ethereum.ledger.ChainReorganised]] - a better new branch was found causing chain reorganisation
-    *         - [[io.iohk.ethereum.ledger.DuplicateBlock]] - block already exists either in the main chain or in the queue
-    *         - [[io.iohk.ethereum.ledger.BlockImportFailed]] - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[BlockImportedToTop]] - if the block was added as the new best block
+    *   - [[BlockEnqueued]]      - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
+    *   - [[ChainReorganised]]   - a better new branch was found causing chain reorganisation
+    *   - [[DuplicateBlock]]     - block already exists either in the main chain or in the queue
+    *   - [[BlockImportFailed]]  - block failed to execute (when importing to top or reorganising the chain)
     */
-  def importBlock(
+  override def evaluateBranchBlock(
       block: Block
   )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[BlockImportResult] =
     blockchainReader.getBestBlock() match {
```

Comment:
I wonder if we can flatten this `match` structure somehow to make this clearer? Although moving all this code out already helps a bunch.

Reply:
I agree with you, still does not look great. With the follow-up ticket ETCM-1069 this will improve :)